### DISCUSSION TITLE
Question about how to split page by using C-API

### DISCUSSION THREAD

**Discussion by mslichao** [qpdf/qpdf#1129] [Q&A] [answered]
Hello qpdf experts,

I have been using qpdf to split a PDF into single-page PDFs with the following command: `qpdf input.pdf --split-pages=1 output.pdf`. Now, I would like to achieve the same result in C# by calling the C-API. However, I would like to use memory instead of the file system for both input and output.

I believe I can use `qpdf_read_memory` to load the input file, but I am not sure how to get each split page PDF in memory. Can anyone guide me on how to do this?

Thank you!

**Comment by jberkenbilt** [comment]
There's no mechanism in qpdf right now to split to memory, but I'll tag this with `pages` and see if I can work into the existing work I'm doing around this category of features.

**Comment by mslichao** [comment]
Thanks @jberkenbilt .
Do you have ETA for the existing work around the pages feature?
If it will take a long time, I may need add extra C-API to support my requirement. This will be a doable action, right?

**Comment by jberkenbilt** [comment, marked as answer]
I'm not sure what your definition of a long time is, but it will be at least weeks, maybe longer. As of qpdf 11.7.0 (latest is 11.8.0), it is possible for you to write your own extern "C" functions in C++ that can interoperate with the C API. Supporting everything the full C++ API does from the C API is a non-goal, so adding your own functions to fill in the gaps is the recommended approach. If you run into trouble with it, let me know. I had played around with a more elaborate C API extension mechanism that included cleanup hooks, but I decided it was too complex and delegated too much responsibility to qpdf. You can find a few files in the examples directory starting with `extend-c-api` that illustrate how you can add your own C functions that interoperate with QPDF's C API.

**Comment by mslichao** [comment]
Yes, I saw the `extend-c-api`, I will customize my C API. Thanks.
